api/git/service.go

```diff
@@ -0,0 +1,155 @@
+package git
+
+import (
+	"context"
+	"errors"
+	"sync"
+	"time"
+)
+
+var (
+	ErrIncorrectRepositoryURL = errors.New("Git repository could not be found, please ensure that the URL is correct.")
+	ErrAuthenticationFailure  = errors.New("Authentication failed, please ensure that the git credentials are correct.")
+	ErrRefNotFound            = errors.New("The target ref is not found")
+
+	REPOSITORY_CACHE_SIZE = 4
+	REPOSITORY_CACHE_TTL  = 30 * time.Minute
+)
+
+type option struct {
+	repositoryUrl string
+	username      string
+	password      string
+	referenceName string
+	depth         int
+	extensions    []string
+}
+
+type repoManager interface {
+	download(ctx context.Context, dst string, opt option) error
+	latestCommitID(ctx context.Context, opt option) (string, error)
+	listRefs(ctx context.Context, opt option) ([]string, error)
+	listFiles(ctx context.Context, opt option) ([]string, error)
+	purgeCache()
+}
+
+// Service represents a service for managing Git.
+type Service struct {
+	shutdownCtx  context.Context
+	azure        repoManager
+	git          repoManager
+	timerStopped bool
+	mut          sync.Mutex
+}
+
+// NewService initializes a new service.
+func NewService(ctx context.Context) *Service {
+	service := &Service{
+		shutdownCtx:  ctx,
+		azure:        NewAzureClient(REPOSITORY_CACHE_SIZE),
+		git:          NewGitClient(REPOSITORY_CACHE_SIZE),
+		timerStopped: false,
+	}
+	go service.startCacheCleanTimer(REPOSITORY_CACHE_TTL)
+	return service
+}
+
+// startCacheCleanTimer starts a timer to purge caches periodically
+func (service *Service) startCacheCleanTimer(d time.Duration) {
```

Comment:
I'm afraid that's not a good way of invalidating a cache because it cleans the whole cache regardless of how recently an item was added. Depending on a particular case, that would make the whole cache useless. That's why a TTL per item is a stronger concept.

Reply:
My idea was that users will play this feature very heavily in a short period and the most of time, it will be untouched. As you said the TTL also consume the CPU. If it's a TTL per item, It will be 8 more TTL at the worst case after moving the cache check into service method. What if later we increase the default cache size? It's just my concern but I'm happy with the TTL per item concept too. 
If we update TTL per item, are you happy with the library https://github.com/projectdiscovery/expirablelru ? or other options?